File: backend/mcp-code-intelligence-nodejs/src/parsers/languages/salesforce-meta-parser.ts
/**
 * KSA-191: Salesforce Metadata Parser.
 * Extracts symbols and relationships from Salesforce metadata XML files.
 * Supports: Flows, Custom Objects, Fields, LWC metadata, Aura metadata.
 * Uses regex-based XML extraction (no tree-sitter — wasmPath is null).
 */

import type {
  ILanguageParser, ParseResult, ExtractedSymbol, ExtractedRelationship,
  ParseError, RelationshipKind
} from '../types.js';

export default class SalesforceMetaParser implements ILanguageParser {
  readonly languageId: string;

  constructor(_parser: any, languageId: string) {
    // _parser is null — this parser does not use tree-sitter
    this.languageId = languageId;
  }

  getSupportedExtensions(): string[] {
    return ['.flow-meta.xml', '.object-meta.xml', '.field-meta.xml',
            '.js-meta.xml', '.component-meta.xml'];
  }

  parse(source: string, filePath: string): ParseResult {
    const symbols: ExtractedSymbol[] = [];
    const relationships: ExtractedRelationship[] = [];
    const errors: ParseError[] = [];

    try {
      const metaType = this.detectMetaType(filePath);
      switch (metaType) {
        case 'flow':
          this.parseFlow(source, filePath, symbols, relationships);
          break;
        case 'object':
          this.parseObject(source, filePath, symbols, relationships);
          break;
        case 'field':
          this.parseField(source, filePath, symbols, relationships);
          break;
        case 'lwc-meta':
          this.parseLWCMeta(source, filePath, symbols, relationships);
          break;
        case 'aura-meta':
          this.parseAuraMeta(source, filePath, symbols);
          break;
        default:
          break;
      }
    } catch (err) {
      errors.push({
        message: `XML parse error: ${err instanceof Error ? err.message : String(err)}`,
        line: 1,
        column: 0,
      });
    }

    return { symbols, relationships, errors };
  }

  // --- Metadata Type Detection ---

  private detectMetaType(filePath: string): string | null {
    const normalized = filePath.replace(/\\/g, '/').toLowerCase();
    if (normalized.endsWith('.flow-meta.xml')) return 'flow';
    if (normalized.endsWith('.object-meta.xml')) return 'object';
    if (normalized.endsWith('.field-meta.xml')) return 'field';
    if (normalized.endsWith('.js-meta.xml')) return 'lwc-meta';
    if (normalized.endsWith('.component-meta.xml')) return 'aura-meta';
    return null;
  }

  // --- Flow Parsing ---

  private parseFlow(
    source: string, filePath: string,
    symbols: ExtractedSymbol[], relationships: ExtractedRelationship[],
  ): void {
    const flowName = this.nameFromPath(filePath);
    const processType = this.extractXmlValues(source, 'processType')[0] ?? 'Flow';
    const lineCount = source.split('\n').length;

    symbols.push({
      name: flowName,
      kind: 'class',
      filePath,
      startLine: 1,
      endLine: lineCount,
      signature: `Flow: ${flowName} (${processType})`,
      modifiers: [processType.toLowerCase()],
      isExported: true,
    });

    // Extract variables as properties
    const variables = this.extractXmlBlocks(source, 'variables');
    for (const varBlock of variables) {
      const varName = this.extractXmlValues(varBlock, 'name')[0];
      const dataType = this.extractXmlValues(varBlock, 'dataType')[0] ?? 'String';
      if (varName) {
        symbols.push({
          name: varName,
          kind: 'property',
          filePath,
          startLine: 1,
          endLine: 1,
          signature: `${varName}: ${dataType}`,
          parentName: flowName,
          returnType: dataType,
          isExported: false,
        });
      }
    }

    // Extract decisions as methods
    const decisions = this.extractXmlBlocks(source, 'decisions');
    for (const block of decisions) {
      const name = this.extractXmlValues(block, 'name')[0];
      if (name) {
        symbols.push({
          name,
          kind: 'method',
          filePath,
          startLine: 1,
          endLine: 1,
          signature: `Decision: ${name}`,
          parentName: flowName,
          isExported: false,
        });
      }
    }

    // Extract actionCalls — Apex invocations
    const actions = this.extractXmlBlocks(source, 'actionCalls');
    for (const block of actions) {
      const actionName = this.extractXmlValues(block, 'name')[0];
      const actionType = this.extractXmlValues(block, 'actionType')[0];
      if (actionName) {
        symbols.push({
          name: actionName,
          kind: 'method',
          filePath,
          startLine: 1,
          endLine: 1,
          signature: `Action: ${actionName} (${actionType ?? 'unknown'})`,
          parentName: flowName,
          isExported: false,
        });
        // If Apex action, create 'calls' relationship
        if (actionType === 'apex') {
          const className = this.extractXmlValues(block, 'actionName')[0];
          if (className) {
            relationships.push({
              sourceSymbol: flowName,
              targetSymbol: className,
              kind: 'calls',
              filePath,
              line: 1,
              metadata: { actionType: 'apex' },
            });
          }
        }
      }
    }

    // Extract referenced SObjects from recordLookups, recordCreates, etc.
    for (const tag of ['recordLookups', 'recordCreates', 'recordUpdates', 'recordDeletes']) {
      const blocks = this.extractXmlBlocks(source, tag);
      for (const block of blocks) {
        const objectName = this.extractXmlValues(block, 'object')[0];
        if (objectName) {
          relationships.push({
            sourceSymbol: flowName,
            targetSymbol: objectName,
            kind: 'uses',
            filePath,
            line: 1,
            metadata: { operation: tag.replace('record', '').toLowerCase() },
          });
        }
      }
    }
  }

  // --- Object Parsing ---

  private parseObject(
    source: string, filePath: string,
    symbols: ExtractedSymbol[], relationships: ExtractedRelationship[],
  ): void {
    const objectName = this.nameFromPath(filePath);
    const lineCount = source.split('\n').length;

    symbols.push({
      name: objectName,
      kind: 'class',
      filePath,
      startLine: 1,
      endLine: lineCount,
      signature: `CustomObject: ${objectName}`,
      modifiers: ['custom-object'],
      isExported: true,
    });

    // Extract fields
    const fields = this.extractXmlBlocks(source, 'fields');
    for (const block of fields) {
      const fieldName = this.extractXmlValues(block, 'fullName')[0];
      const fieldType = this.extractXmlValues(block, 'type')[0] ?? 'Text';
      if (fieldName) {
        symbols.push({
          name: fieldName,
          kind: 'property',
          filePath,
          startLine: 1,
          endLine: 1,
          signature: `${fieldName}: ${fieldType}`,
          parentName: objectName,
          returnType: fieldType,
          isExported: true,
        });

        // Lookup/MasterDetail relationships
        if (fieldType === 'Lookup' || fieldType === 'MasterDetail') {
          const referenceTo = this.extractXmlValues(block, 'referenceTo')[0];
          if (referenceTo) {
            relationships.push({
              sourceSymbol: objectName,
              targetSymbol: referenceTo,
              kind: 'uses',
              filePath,
              line: 1,
              metadata: { relationType: fieldType },
            });
          }
        }
      }
    }

    // Extract validation rules
    const validations = this.extractXmlBlocks(source, 'validationRules');
    for (const block of validations) {
      const ruleName = this.extractXmlValues(block, 'fullName')[0];
      if (ruleName) {
        symbols.push({
          name: ruleName,
          kind: 'method',
          filePath,
          startLine: 1,
          endLine: 1,
          signature: `ValidationRule: ${ruleName}`,
          parentName: objectName,
          isExported: false,
        });
      }
    }
  }

  // --- Field Parsing (standalone .field-meta.xml) ---

  private parseField(
    source: string, filePath: string,
    symbols: ExtractedSymbol[], relationships: ExtractedRelationship[],
  ): void {
    const fieldName = this.nameFromPath(filePath);
    const fieldType = this.extractXmlValues(source, 'type')[0] ?? 'Text';
    const parentObject = this.inferObjectFromFieldPath(filePath);
    const lineCount = source.split('\n').length;

    symbols.push({
      name: fieldName,
      kind: 'property',
      filePath,
      startLine: 1,
      endLine: lineCount,
      signature: `${fieldName}: ${fieldType}`,
      parentName: parentObject,
      returnType: fieldType,
      isExported: true,
    });

    if (fieldType === 'Lookup' || fieldType === 'MasterDetail') {
      const referenceTo = this.extractXmlValues(source, 'referenceTo')[0];
      if (referenceTo && parentObject) {
        relationships.push({
          sourceSymbol: parentObject,
          targetSymbol: referenceTo,
          kind: 'uses',
          filePath,
          line: 1,
          metadata: { relationType: fieldType, field: fieldName },
        });
      }
    }
  }

  // --- LWC Metadata Parsing ---

  private parseLWCMeta(
    source: string, filePath: string,
    symbols: ExtractedSymbol[], relationships: ExtractedRelationship[],
  ): void {
    const componentName = this.nameFromPath(filePath);
    const lineCount = source.split('\n').length;

    // Check if exposed
    const isExposed = this.extractXmlValues(source, 'isExposed')[0] === 'true';

    // Extract targets
    const targets = this.extractXmlValues(source, 'target');

    symbols.push({
      name: componentName,
      kind: 'class',
      filePath,
      startLine: 1,
      endLine: lineCount,
      signature: `LWC: ${componentName}`,
      modifiers: isExposed ? ['exposed'] : [],
      isExported: isExposed,
    });

    // Extract wire adapters — @wire relationships
    const wireBlocks = this.extractXmlBlocks(source, 'targetConfig');
    // Also check for datasource references
    const datasources = this.extractXmlValues(source, 'datasource');
    for (const ds of datasources) {
      if (ds) {
        relationships.push({
          sourceSymbol: componentName,
          targetSymbol: ds,
          kind: 'wire' as RelationshipKind,
          filePath,
          line: 1,
          metadata: { type: 'datasource' },
        });
      }
    }
  }

  // --- Aura Metadata Parsing ---

  private parseAuraMeta(
    source: string, filePath: string,
    symbols: ExtractedSymbol[],
  ): void {
    const componentName = this.nameFromPath(filePath);
    const lineCount = source.split('\n').length;

    symbols.push({
      name: componentName,
      kind: 'class',
      filePath,
      startLine: 1,
      endLine: lineCount,
      signature: `AuraComponent: ${componentName}`,
      modifiers: ['aura'],
      isExported: true,
    });
  }

  // --- XML Extraction Helpers ---

  /** Extract XML element text content by tag name. */
  private extractXmlValues(source: string, tagName: string): string[] {
    const regex = new RegExp(`<${tagName}>([^<]*)</${tagName}>`, 'g');
    const results: string[] = [];
    let match: RegExpExecArray | null;
    while ((match = regex.exec(source)) !== null) {
      results.push(match[1]);
    }
    return results;
  }

  /** Extract XML blocks (multi-line elements). */
  private extractXmlBlocks(source: string, tagName: string): string[] {
    const regex = new RegExp(`<${tagName}>[\\s\\S]*?</${tagName}>`, 'g');
    const results: string[] = [];
    let match: RegExpExecArray | null;
    while ((match = regex.exec(source)) !== null) {
      results.push(match[0]);
    }
    return results;
  }

  /** Extract component name from file path. */
  private nameFromPath(filePath: string): string {
    const normalized = filePath.replace(/\\/g, '/');
    const basename = normalized.split('/').pop() ?? normalized;
    // Remove all meta extensions: .flow-meta.xml -> name
    return basename
      .replace(/\.(flow|object|field|js|component)-meta\.xml$/, '')
      .replace(/\.\w+$/, '');
  }

  /** Infer parent object from field path. */
  private inferObjectFromFieldPath(filePath: string): string | null {
    const normalized = filePath.replace(/\\/g, '/');
    // Pattern: .../objects/{ObjectName}/fields/{FieldName}.field-meta.xml
    const match = normalized.match(/objects\/([^/]+)\/fields\//);
    return match ? match[1] : null;
  }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/parsers/languages/salesforce_meta_parser.py
"""
KSA-191: Salesforce Metadata Parser.
Extracts symbols and relationships from Salesforce metadata XML files:
  - .flow-meta.xml (Flows)
  - .object-meta.xml (Custom Objects)
  - .field-meta.xml (Custom Fields)
  - .js-meta.xml (LWC metadata)
  - .component-meta.xml (Aura metadata)

Uses regex-based XML extraction (no tree-sitter — XML files).
Constructor receives None as parser since no WASM grammar is needed.
"""

from __future__ import annotations

import re
from typing import Any

from ..types import (
    ExtractedRelationship, ExtractedSymbol, ParseError, ParseResult,
)


class SalesforceMetaParser:
    """Extracts symbols and relationships from Salesforce metadata XML files."""

    def __init__(self, parser: Any, language_id: str) -> None:
        # parser is None — this parser doesn't use tree-sitter
        self._language_id = language_id

    @property
    def language_id(self) -> str:
        return self._language_id

    def get_supported_extensions(self) -> list[str]:
        return [
            ".flow-meta.xml", ".object-meta.xml", ".field-meta.xml",
            ".js-meta.xml", ".component-meta.xml",
        ]

    def parse(self, source: str, file_path: str) -> ParseResult:
        symbols: list[ExtractedSymbol] = []
        relationships: list[ExtractedRelationship] = []
        errors: list[ParseError] = []

        try:
            meta_type = self._detect_meta_type(file_path)
            if meta_type == "flow":
                self._parse_flow(source, file_path, symbols, relationships)
            elif meta_type == "object":
                self._parse_object(source, file_path, symbols, relationships)
            elif meta_type == "field":
                self._parse_field(source, file_path, symbols, relationships)
            elif meta_type == "lwc-meta":
                self._parse_lwc_meta(source, file_path, symbols, relationships)
            elif meta_type == "aura-meta":
                self._parse_aura_meta(source, file_path, symbols)
        except Exception as err:
            errors.append(ParseError(
                message=f"XML parse error: {err}",
                line=1, column=0,
            ))

        return ParseResult(symbols=symbols, relationships=relationships, errors=errors)

    # --- Metadata type detection ---

    def _detect_meta_type(self, file_path: str) -> str | None:
        """Detect metadata type from file extension."""
        lower = file_path.lower()
        if lower.endswith(".flow-meta.xml"):
            return "flow"
        if lower.endswith(".object-meta.xml"):
            return "object"
        if lower.endswith(".field-meta.xml"):
            return "field"
        if lower.endswith(".js-meta.xml"):
            return "lwc-meta"
        if lower.endswith(".component-meta.xml"):
            return "aura-meta"
        return None

    # --- XML parsing helpers (regex-based) ---

    def _extract_xml_values(self, source: str, tag_name: str) -> list[str]:
        """Extract XML element text content by tag name."""
        pattern = re.compile(rf"<{re.escape(tag_name)}>([^<]*)</{re.escape(tag_name)}>")
        return pattern.findall(source)

    def _extract_xml_blocks(self, source: str, tag_name: str) -> list[str]:
        """Extract XML blocks (multi-line elements)."""
        pattern = re.compile(
            rf"<{re.escape(tag_name)}>([\s\S]*?)</{re.escape(tag_name)}>", re.DOTALL
        )
        return [m.group(0) for m in pattern.finditer(source)]

    # --- Flow parsing ---

    def _parse_flow(
        self, source: str, file_path: str,
        symbols: list[ExtractedSymbol], relationships: list[ExtractedRelationship],
    ) -> None:
        """Parse .flow-meta.xml — extract flow structure, variables, decisions, actions."""
        flow_name = self._name_from_path(file_path)
        process_type = self._extract_xml_values(source, "processType")
        process_type_str = process_type[0] if process_type else "Flow"
        line_count = source.count("\n") + 1

        symbols.append(ExtractedSymbol(
            name=flow_name, kind="class", file_path=file_path,
            start_line=1, end_line=line_count,
            signature=f"Flow: {flow_name} ({process_type_str})",
            modifiers=[process_type_str.lower()], is_exported=True,
        ))

        # Extract variables as properties
        variables = self._extract_xml_blocks(source, "variables")
        for var_block in variables:
            var_name_list = self._extract_xml_values(var_block, "name")
            data_type_list = self._extract_xml_values(var_block, "dataType")
            var_name = var_name_list[0] if var_name_list else None
            data_type = data_type_list[0] if data_type_list else "String"
            if var_name:
                symbols.append(ExtractedSymbol(
                    name=var_name, kind="property", file_path=file_path,
                    start_line=1, end_line=1,
                    signature=f"{var_name}: {data_type}",
                    parent_name=flow_name, return_type=data_type,
                    is_exported=False,
                ))

        # Extract decisions as methods
        decisions = self._extract_xml_blocks(source, "decisions")
        for block in decisions:
            name_list = self._extract_xml_values(block, "name")
            name = name_list[0] if name_list else None
            if name:
                symbols.append(ExtractedSymbol(
                    name=name, kind="method", file_path=file_path,
                    start_line=1, end_line=1,
                    signature=f"Decision: {name}",
                    parent_name=flow_name, is_exported=False,
                ))

        # Extract actionCalls — Apex invocations
        actions = self._extract_xml_blocks(source, "actionCalls")
        for block in actions:
            action_name_list = self._extract_xml_values(block, "name")
            action_type_list = self._extract_xml_values(block, "actionType")
            action_name = action_name_list[0] if action_name_list else None
            action_type = action_type_list[0] if action_type_list else None
            if action_name:
                symbols.append(ExtractedSymbol(
                    name=action_name, kind="method", file_path=file_path,
                    start_line=1, end_line=1,
                    signature=f"Action: {action_name} ({action_type})",
                    parent_name=flow_name, is_exported=False,
                ))
                # If Apex action, create 'calls' relationship
                if action_type == "apex":
                    class_name_list = self._extract_xml_values(block, "actionName")
                    class_name = class_name_list[0] if class_name_list else None
                    if class_name:
                        relationships.append(ExtractedRelationship(
                            source_symbol=flow_name, target_symbol=class_name,
                            kind="calls", file_path=file_path,
                            line=1, metadata={"actionType": "apex"},
                        ))

        # Extract referenced SObjects from record operations
        for tag in ("recordLookups", "recordCreates", "recordUpdates", "recordDeletes"):
            blocks = self._extract_xml_blocks(source, tag)
            for block in blocks:
                object_name_list = self._extract_xml_values(block, "object")
                object_name = object_name_list[0] if object_name_list else None
                if object_name:
                    relationships.append(ExtractedRelationship(
                        source_symbol=flow_name, target_symbol=object_name,
                        kind="uses", file_path=file_path,
                        line=1, metadata={"operation": tag.replace("record", "").lower()},
                    ))

    # --- Object parsing ---

    def _parse_object(
        self, source: str, file_path: str,
        symbols: list[ExtractedSymbol], relationships: list[ExtractedRelationship],
    ) -> None:
        """Parse .object-meta.xml — extract object definition, fields, validation rules."""
        object_name = self._name_from_path(file_path)
        line_count = source.count("\n") + 1

        symbols.append(ExtractedSymbol(
            name=object_name, kind="class", file_path=file_path,
            start_line=1, end_line=line_count,
            signature=f"CustomObject: {object_name}",
            modifiers=["custom-object"], is_exported=True,
        ))

        # Extract fields
        fields = self._extract_xml_blocks(source, "fields")
        for block in fields:
            field_name_list = self._extract_xml_values(block, "fullName")
            field_type_list = self._extract_xml_values(block, "type")
            field_name = field_name_list[0] if field_name_list else None
            field_type = field_type_list[0] if field_type_list else "Text"
            if field_name:
                symbols.append(ExtractedSymbol(
                    name=field_name, kind="property", file_path=file_path,
                    start_line=1, end_line=1,
                    signature=f"{field_name}: {field_type}",
                    parent_name=object_name, return_type=field_type,
                    is_exported=True,
                ))

                # Lookup/MasterDetail relationships
                if field_type in ("Lookup", "MasterDetail"):
                    reference_to_list = self._extract_xml_values(block, "referenceTo")
                    reference_to = reference_to_list[0] if reference_to_list else None
                    if reference_to:
                        relationships.append(ExtractedRelationship(
                            source_symbol=object_name, target_symbol=reference_to,
                            kind="uses", file_path=file_path,
                            line=1, metadata={"relationType": field_type},
                        ))

        # Extract validation rules
        validations = self._extract_xml_blocks(source, "validationRules")
        for block in validations:
            rule_name_list = self._extract_xml_values(block, "fullName")
            rule_name = rule_name_list[0] if rule_name_list else None
            if rule_name:
                symbols.append(ExtractedSymbol(
                    name=rule_name, kind="method", file_path=file_path,
                    start_line=1, end_line=1,
                    signature=f"ValidationRule: {rule_name}",
                    parent_name=object_name, is_exported=False,
                ))

    # --- Field parsing (standalone .field-meta.xml) ---

    def _parse_field(
        self, source: str, file_path: str,
        symbols: list[ExtractedSymbol], relationships: list[ExtractedRelationship],
    ) -> None:
        """Parse .field-meta.xml — standalone field definition."""
        field_name = self._name_from_path(file_path)
        field_type_list = self._extract_xml_values(source, "type")
        field_type = field_type_list[0] if field_type_list else "Text"
        parent_object = self._infer_object_from_field_path(file_path)
        line_count = source.count("\n") + 1

        symbols.append(ExtractedSymbol(
            name=field_name, kind="property", file_path=file_path,
            start_line=1, end_line=line_count,
            signature=f"{field_name}: {field_type}",
            parent_name=parent_object, return_type=field_type,
            is_exported=True,
        ))

        if field_type in ("Lookup", "MasterDetail"):
            reference_to_list = self._extract_xml_values(source, "referenceTo")
            reference_to = reference_to_list[0] if reference_to_list else None
            if reference_to and parent_object:
                relationships.append(ExtractedRelationship(
                    source_symbol=parent_object, target_symbol=reference_to,
                    kind="uses", file_path=file_path,
                    line=1, metadata={"relationType": field_type, "field": field_name},
                ))

    # --- LWC metadata parsing ---

    def _parse_lwc_meta(
        self, source: str, file_path: str,
        symbols: list[ExtractedSymbol], relationships: list[ExtractedRelationship],
    ) -> None:
        """Parse .js-meta.xml — LWC component metadata."""
        component_name = self._name_from_path(file_path)
        line_count = source.count("\n") + 1

        # Check if exposed
        is_exposed_list = self._extract_xml_values(source, "isExposed")
        is_exposed = is_exposed_list[0].lower() == "true" if is_exposed_list else False

        # Extract targets
        targets = self._extract_xml_values(source, "target")

        symbols.append(ExtractedSymbol(
            name=component_name, kind="class", file_path=file_path,
            start_line=1, end_line=line_count,
            signature=f"LWC: {component_name}",
            modifiers=["lwc"] + (["exposed"] if is_exposed else []),
            is_exported=is_exposed,
        ))

        # Extract @wire targets from targetConfigs
        wire_targets = self._extract_xml_values(source, "objects")
        for obj in wire_targets:
            relationships.append(ExtractedRelationship(
                source_symbol=component_name, target_symbol=obj,
                kind="wire", file_path=file_path,
                line=1, metadata={"targets": targets},
            ))

    # --- Aura metadata parsing ---

    def _parse_aura_meta(
        self, source: str, file_path: str,
        symbols: list[ExtractedSymbol],
    ) -> None:
        """Parse .component-meta.xml — Aura component metadata (minimal extraction)."""
        component_name = self._name_from_path(file_path)
        line_count = source.count("\n") + 1

        # Extract description
        desc_list = self._extract_xml_values(source, "description")
        description = desc_list[0] if desc_list else None

        symbols.append(ExtractedSymbol(
            name=component_name, kind="class", file_path=file_path,
            start_line=1, end_line=line_count,
            signature=f"AuraComponent: {component_name}",
            modifiers=["aura"], is_exported=True,
            doc_comment=description,
        ))

    # --- Helper methods ---

    def _name_from_path(self, file_path: str) -> str:
        """Extract component name from file path."""
        # Normalize path separators
        normalized = file_path.replace("\\", "/")
        basename = normalized.split("/")[-1] if "/" in normalized else normalized
        # Remove all meta extensions: .flow-meta.xml -> name
        basename = re.sub(
            r"\.(flow|object|field|js|component)-meta\.xml$", "", basename, flags=re.IGNORECASE
        )
        # Remove any remaining extension
        basename = re.sub(r"\.\w+$", "", basename)
        return basename

    def _infer_object_from_field_path(self, file_path: str) -> str | None:
        """Infer parent object from field path.
        Pattern: .../objects/{ObjectName}/fields/{FieldName}.field-meta.xml
        """
        normalized = file_path.replace("\\", "/")
        match = re.search(r"objects/([^/]+)/fields/", normalized)
        return match.group(1) if match else None


# Module-level alias for dynamic loading
Parser = SalesforceMetaParser
